src/common/tensor_inspector.h

```diff
@@ -0,0 +1,733 @@
+  TensorInspector(const mshadow::Tensor<Device, dimension, DType>& ts, const RunContext& ctx):
+      tb_(ts), ctx_(ctx) {}
+
+  /*!
+   * \brief construct from TBlob object
+   * \param tb the source tblob object
+   * \param ctx the run context of the tensor
+   */
+  TensorInspector(const TBlob& tb, const RunContext& ctx):
+      tb_(tb), ctx_(ctx) {}
+
+  /*!
+   * \brief construct from NDArray object. Currently this only works with kDefaultStorage
+   * \param arr the source ndarray object
+   * \param ctx the run context of the tensor
+   */
+  TensorInspector(const NDArray& arr, const RunContext& ctx):
+      tb_(arr.data()), ctx_(ctx) {}
+
+  /*!
+   * \brief print the tensor to std::cout
+   */
+  inline void print_string() {
+    std::cout << to_string() << std::endl;
+  }
+
+  /*!
+   * \brief return a string which contains the values and other info of the tensor
+   */
+  inline std::string to_string() {
+    std::stringstream ss;
+    MSHADOW_TYPE_SWITCH(tb_.type_flag_, DType, {
+      to_string_helper<DType>(&ss);
+    });
+    return ss.str();
+  }
+
+  /*!
+   * \brief interactively print the tensor value
+   * \param tag the name given to this call
+   */
+  inline void interactive_print(std::string tag = "") {
+    MSHADOW_TYPE_SWITCH(tb_.type_flag_, DType, {
+      interactive_print_helper<DType>(tag);
+    });
+  }
+
+  /*!
+   * \brief check/validate the values within the tensor, return the coordinates
+   * where the value checker evaluates to true
+   * \tparam ValueChecker the type of the lambda
+   * \param checker the lambda function to check each value of within the tensor
+   * \param interactive wherether to allow the user to interactively check the coordinates
+   * \param tag the name given to this call
+   */
+  template<typename ValueChecker>
+  inline std::vector<std::vector<int>> check_value(const ValueChecker& checker,
+      bool interactive = false, std::string tag = "") {
+    std::vector<std::vector<int>> ret;
+    MSHADOW_TYPE_SWITCH(tb_.type_flag_, DType, {
+      check_value_helper<DType>(&ret, checker, ret, interactive, tag);
+    });
+    return ret;
+  }
+
+  /*!
+   * \brief check/validate the values within the tensor, return the coordinates
+   * where the lambda evaluates to true
+   * \param ct the type of the checker
+   * \param interactive wherether to allow the user to interactively check the coordinates
+   * \param tag the name given to this call
+   */
+  inline std::vector<std::vector<int>> check_value(CheckerType ct, bool interactive = false,
+      std::string tag = "") {
+    std::vector<std::vector<int>> ret;
+    MSHADOW_TYPE_SWITCH(tb_.type_flag_, DType, {
+      check_value_helper<DType>(&ret, build_checker<DType>(ct), interactive, tag);
+    });
+    return ret;
+  }
+
+  /*!
+   * \brief dump the value of the tensor to a file with name "tag_[visit count].npy" in npy format
+   * \param tag the name given to this call
+   */
+  inline void dump_value(std::string tag) {
```

Comment:
Could we refine the name to dump to file or similar? is not clear where are we dumping to from the name of the function.

Reply:
changed